### PR TITLE
Enable unidoc for all Delta projects that are going to be released to Maven

### PR DESCRIPTION
**Pull request by tdas**
#### Which Delta project/connector is this regarding?
- [x] Spark
- [x] Standalone
- [x] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description
This is leftover task from the connector code migration to this repo. The unidoc/java doc generation was disabled for standalone, and flink. This is to enable them back. Specifically I did the following
1. Refactored all the unidoc source file filtering code in a clean way so that each subproject can configure but reuse all the code. 
2. With this refactoring we can also easily generate combined docs for multiple projects. For example, the 2 Kernel projects can be documented together using the kernelGroup aggregation project. This is what I plan to upload to docs.delta.io. 
3. ~The generated doc for Maven upload is also fixed to make sure we dont public doc jars exposing inner classes. As a result, each project with public APIs will generate it own scala/java docs.~ I reverted this as this is throwing errors that I cannot fix yet. Added a TODO in the code. Also in a separate PR, I will update the higher level doc generation scripts in delta/docs to consolidate the separate docs into one location.

You can generate all the docs by `build/sbt unidoc`

## How was this patch tested?
Existing tests force generation of docs

## Does this PR introduce _any_ user-facing changes?
No

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -69,17 +70,19 @@ lazy val commonSettings = Seq(
...
   testOptions += Tests.Argument("-oF"),
+
+  // Unidoc settings: by default dont document any source file
+  Compile / unidoc / sourceFilePatternsToDocument := Nil,
 )
 
 lazy val spark = (project in file("spark"))
   .dependsOn(storage)
-  .enablePlugins(GenJavadocPlugin, JavaUnidocPlugin, ScalaUnidocPlugin, Antlr4Plugin)
+  .enablePlugins(Antlr4Plugin)
```
Comment (tdas) [inline build.sbt:80]: this has moved in the Unidoc.scala below.

File: build.sbt [hunk 2]
```diff
@@ -923,7 +939,7 @@ lazy val flink = (project in file("connectors/flink"))
     // generating source java class with version number to be passed during commit to the DeltaLog as engine info
     // (part of transaction's metadata)
     Compile / sourceGenerators += Def.task {
-      val file = (Compile / sourceManaged).value / "meta" / "Meta.java"
+      val file = (Compile / sourceManaged).value / "io" / "delta" / "flink" / "internal" / "Meta.java"
```
Comment (tdas) [inline build.sbt:942]: this change is to make sure that this file is inside the internal directory and gets removed during doc generation.

File: build.sbt [hunk 3]
```diff
@@ -987,8 +997,12 @@ lazy val kernelGroup = project
     // crossScalaVersions must be set to Nil on the aggregating project
     crossScalaVersions := Nil,
     publishArtifact := false,
-    publish / skip := false
-  )
+    publish / skip := false,
+    unidocSourceFilePatterns := {
```
Comment (tdas) [inline build.sbt:1001]: this will generate a unified kernel api and default API docs that we will upload to our website.

**Review by allisonport-db — APPROVED** [review]
LGTM (I built them locally as well and worked for me, they all looked right)

**Comment by allisonport-db** [comment]
Seems like it's failing in the spark github actions though? Seems like it's trying to build the docs for all the projects, should it only be per project?

**Comment by tdas** [comment]
The problem is with build/sbt publishM2 which generate and publishes the docs maven artifact, where unidoc is failing for some reason. 

And we use publishM2 to locally publish the packages for running python tests. 

Still debugging.

File: build.sbt [hunk 4]
```diff
@@ -236,14 +242,18 @@ lazy val kernelDefault = (project in file("kernel/kernel-default"))
...
     // and during tests (e.g. build/sbt test)
     checkstyleConfigLocation := CheckstyleConfigLocation.File("kernel/dev/checkstyle.xml"),
     checkstyleSeverityLevel := Some(CheckstyleSeverityLevel.Error),
     (Compile / checkstyle) := (Compile / checkstyle).triggeredBy(Compile / compile).value,
-    (Test / checkstyle) := (Test / checkstyle).triggeredBy(Test / compile).value
-  )
+    (Test / checkstyle) := (Test / checkstyle).triggeredBy(Test / compile).value,
+
+      // Unidoc settings
+    unidocSourceFilePatterns += SourceFilePattern("io/delta/kernel/"),
```
[the remark this answers is not in the document]
Comment (tdas) [reply build.sbt:255]: if everything that needs to be hidden is in an "internal" subdir, then everything that needs to public will be in io/delta/kernel subdir. so we dont really need to change anything if we follow that "internal" rule.

**Review by scottsand-db — COMMENTED** [review]
LGTM

**Pull request closed without merging**